backend/src/api/authentication/authentication.gateway.ts

```diff
@@ -0,0 +1,108 @@
+import {
+  OnGatewayConnection,
+  OnGatewayDisconnect,
+  SubscribeMessage,
+  WebSocketGateway,
+  WebSocketServer,
+} from "@nestjs/websockets";
+import { Server, Socket } from "socket.io";
+import { Public } from "./role.decorator";
+import { AuthenticationService } from "./authentication.service";
+
+type Client = Socket;
+
+@WebSocketGateway({
+  // this cannot be configured using dependency injection
+  // see https://github.com/nestjs/nest/issues/7649#issuecomment-964873444
+  cors: true,
+  transports: ["websocket"],
+})
+export class AuthenticationGateway
+  implements OnGatewayConnection, OnGatewayDisconnect
+{
+  readonly socketsByUserId = new Map<number, Client[]>();
+
+  @WebSocketServer() io!: Server;
+
+  constructor(private authenticationService: AuthenticationService) {}
+
+  async handleConnection(client: Client): Promise<void> {
+    const token = client.handshake.auth.token;
+
+    if (!token) {
+      return;
+    }
+
+    const user =
+      await this.authenticationService.findByAuthenticationTokenOrThrow(token);
+
+    if (user && !this.authenticationService.isStudent(user)) {
+      const userId = user.id;
+      if (!this.socketsByUserId.has(userId)) {
+        this.socketsByUserId.set(userId, []);
+      }
+
+      this.socketsByUserId.get(userId)!.push(client);
+    }
+  }
+
+  handleDisconnect(client: Client): void {
+    const { user } = client.data;
+
+    if (user) {
+      const sockets = this.socketsByUserId.get(user.id);
+      if (!sockets) {
+        return;
+      }
+
+      const newSockets = sockets.filter((s) => s.id !== client.id);
+
+      if (newSockets.length === 0) {
+        this.socketsByUserId.delete(user.id);
+      } else {
+        this.socketsByUserId.set(user.id, newSockets);
+      }
+    }
+  }
+
+  @SubscribeMessage("requestTeacherToSignInStudent")
+  @Public()
+  requestTeacherToSignInStudent(
+    client: Client,
+    payload: { teacherId: number },
+  ): void {
+    const { teacherId, ...rest } = payload;
+    const teacherSockets = this.socketsByUserId.get(teacherId);
+
+    if (!teacherSockets) {
+      throw new Error("Teacher not online");
```

Comment:
Yes it may be interesting, however the client will just keep sending requests making a reply redundant. So while there is no need for it to function correctly, I definitely see your point. I would create a ticket for the future when we do some cleanup and also define the final v1 protocol / websocket API?

Reply:
Yup, please create a ticket so that we can display an appropriate warning message to the user :)

Reply:
https://classroomrt.atlassian.net/browse/CRT-50